File: Design-Patterns-Using-Go/Proxy/ProxyLazy.go
package main

import "fmt"

// Subject is an interface representing a subject.
type Subject interface {
	Request()
}

// RealSubject is a concrete implementation of the Subject interface.
type RealSubject struct{}

func (rs *RealSubject) Request() {
	fmt.Println("Concrete Subject Request Method")
}

// Proxy is a proxy implementation of the Subject interface.
type Proxy struct {
	subject Subject
}

func (p *Proxy) Request() {
	if p.subject == nil {
		p.subject = &RealSubject{} // Lazy Init
	}
	p.subject.Request()
}

// Client code
func main() {
	proxy := &Proxy{}
	proxy.Request()
}

/*
Concrete Subject Request Method
*/

File: Design-Patterns-Using-Cpp/Proxy/ProxyLazy.cpp
#include <iostream>

class Subject {
public:
    virtual void request() = 0;
};

class RealSubject : public Subject {
public:
    void request() override {
        std::cout << "Concrete Subject Request Method" << std::endl;
    }
};

class Proxy : public Subject {
private:
    RealSubject* realSubject;

public:
    Proxy() : realSubject(nullptr) {}

    void request() override {
        if (realSubject == nullptr) {
            realSubject = new RealSubject(); // Lazy Initialization
        }
        realSubject->request();
    }

    ~Proxy() {
        delete realSubject; // Clean up allocated memory
    }
};

// Client code
int main() {
    Proxy proxy;
    proxy.request();

    return 0;
}

/*
Concrete Subject Request Method
*/